src/iam/certhandler/modules/certmodule.cpp

```diff
@@ -0,0 +1,418 @@
+    }
+
+    err = mX509Provider->CreateCertificate(*templ, *templ, *key.mValue, *pemCert);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return ApplyCert(*pemCert, *certInfo);
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+Error CertModule::RemoveInvalidCerts(const String& password)
+{
+    for (const auto& url : mInvalidCerts) {
+        LOG_WRN() << "Remove invalid cert. type=" << GetCertType() << ", url=" << url;
+
+        const auto err = mHSM->RemoveCert(url, password);
+        if (!err.IsNone()) {
+            return err;
+        }
+    }
+
+    mInvalidCerts.Clear();
+
+    return Error::Enum::eNone;
+}
+
+Error CertModule::RemoveInvalidKeys(const String& password)
+{
+    for (const auto& url : mInvalidKeys) {
+        LOG_WRN() << "Remove invalid key: type=" << GetCertType() << ", url=" << url;
+
+        const auto err = mHSM->RemoveKey(url, password);
+        if (!err.IsNone()) {
+            return err;
+        }
+    }
+
+    mInvalidKeys.Clear();
+
+    return Error::Enum::eNone;
+}
+
+Error CertModule::TrimCerts(const String& password)
+{
+    auto certsInStorage = MakeUnique<StorageCerts>(&mAllocator);
+    if (!certsInStorage) {
+        return AOS_ERROR_WRAP(Error::Enum::eNoMemory);
+    }
+
+    auto err = mStorage->GetCertsInfo(GetCertType(), *certsInStorage);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (certsInStorage->Size() > mModuleConfig.mMaxCertificates) {
+        LOG_WRN() << "Current cert count exceeds max count: " << certsInStorage->Size() << " > "
+                  << mModuleConfig.mMaxCertificates << ". Remove old certificates";
+    }
+
+    while (certsInStorage->Size() > mModuleConfig.mMaxCertificates) {
+        time::Time minTime;
+        CertInfo*  info = nullptr;
+
+        for (auto& cert : *certsInStorage) {
+            if (minTime.IsZero() || cert.mNotAfter < minTime) {
+                minTime = cert.mNotAfter;
+                info = &cert;
+            }
+        }
+
+        err = mHSM->RemoveCert(info->mCertURL, password);
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        err = mHSM->RemoveKey(info->mKeyURL, password);
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        err = mStorage->RemoveCertInfo(GetCertType(), info->mCertURL);
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        certsInStorage->Remove(info);
+    }
+
+    return Error::Enum::eNone;
+}
+
+Error CertModule::CheckCertChain(const Array<crypto::x509::Certificate>& chain)
+{
+    if (chain.IsEmpty()) {
+        return AOS_ERROR_WRAP(Error::Enum::eNotFound);
+    }
+
+    for (const auto& cert : chain) {
+        auto issuer = MakeUnique<DNString>(&mAllocator);
```

Comment:
DNString is ok to keep in stack.

Reply:
For now arrays with certificates and pem blobs will be stored in the allocator.